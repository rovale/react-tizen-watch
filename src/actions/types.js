export const INITIALIZE_APP = 'INITIALIZE_APP';

export const LOAD_PAGES = 'LOAD_PAGES';
export const LOAD_DEVICES = 'LOAD_DEVICES';
export const LOAD_GROUPS = 'LOAD_GROUPS';

export const ACTIVATE_PAGE = 'ACTIVATE_PAGE';
export const SELECT_PAGE = 'SELECT_PAGE';
export const CLOSE_PAGE = 'CLOSE_PAGE';

export const ACTIVATE_DEVICE = 'ACTIVATE_DEVICE';
export const SELECT_DEVICE = 'SELECT_DEVICE';
export const CLOSE_DEVICE = 'CLOSE_DEVICE';
export const CHANGE_DEVICE_ATTRIBUTE = 'CHANGE_DEVICE_ATTRIBUTE';
export const TOGGLE_SWITCH = 'TOGGLE_SWITCH';

export const ACTIVATE_BUTTON = 'ACTIVATE_BUTTON';
export const CLICK_BUTTON = 'CLICK_BUTTON';
