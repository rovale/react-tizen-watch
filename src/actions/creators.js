import * as actionType from './types';

export const initializeApp = () => ({ type: actionType.INITIALIZE_APP });
export const handleError = (msg, url, lineNo, columnNo, error) => ({ type: actionType.HANDLE_ERROR, payload: { msg, url, lineNo, columnNo, error } });
export const loadPages = pages => ({ type: actionType.LOAD_PAGES, payload: { pages } });
export const loadDevices = devices => ({ type: actionType.LOAD_DEVICES, payload: { devices } });

export const activatePage = id => ({ type: actionType.ACTIVATE_PAGE, payload: { id } });
export const selectPage = id => ({ type: actionType.SELECT_PAGE, payload: { id } });
export const closePage = () => ({ type: actionType.CLOSE_PAGE });

export const activateDevice = id => ({ type: actionType.ACTIVATE_DEVICE, payload: { id } });
export const selectDevice = id => ({ type: actionType.SELECT_DEVICE, payload: { id } });
export const closeDevice = () => ({ type: actionType.CLOSE_DEVICE });
export const changeDeviceAttribute = event => ({ type: actionType.CHANGE_DEVICE_ATTRIBUTE, payload: { event } });

export const toggleSwitch = id => ({ type: actionType.TOGGLE_SWITCH, payload: { id } });

export const activateButton = id => ({ type: actionType.ACTIVATE_BUTTON, payload: { id } });
export const clickButton = id => ({ type: actionType.CLICK_BUTTON, payload: { id } });
