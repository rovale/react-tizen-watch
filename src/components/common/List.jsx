import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

export class List extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.object).isRequired,
    activeOptionId: PropTypes.string.isRequired,
    onActivateOption: PropTypes.func.isRequired,
    onSelectOption: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);

    this.onRotaryDetent = this.onRotaryDetent.bind(this);
    this.onSelectItem = this.onSelectItem.bind(this);
  }

  componentDidMount() {
    window.addEventListener('rotarydetent', this.onRotaryDetent);
  }

  componentWillUnmount() {
    window.removeEventListener('rotarydetent', this.onRotaryDetent);
  }

  onRotaryDetent(e) {
    const { direction } = e.detail;

    const index = this.props.options.findIndex(item => item.id === this.props.activeOptionId);

    if (direction === 'CW') {
      if (index < (this.props.options.length - 1)) {
        this.props.onActivateOption(this.props.options[index + 1].id);
      }
    } else if (index > 0) {
      this.props.onActivateOption(this.props.options[index - 1].id);
    }
  }

  onSelectItem(itemId) {
    this.props.onSelectOption(itemId);
  }

  render() {
    const { options } = this.props;

    const items = options.map(item => (
      <Item
        key={item.id}
        id={item.id}
        active={item.id === this.props.activeOptionId}
        onSelect={item.onSelect || this.onSelectItem}
      >
        {item.content || <div>{item.title}</div>}
      </Item>
    ));

    return (
      <ul className="ui-listview ui-snap-listview">
        {items}
      </ul>
    );
  }
}

export class Item extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    onSelect: PropTypes.func.isRequired,
    children: PropTypes.element.isRequired,
  };

  static contextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.scrollToMiddle = this.scrollToMiddle.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.element = null;
  }

  componentDidMount() {
    if (this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(1), 0);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.active && !this.props.active) {
      window.setTimeout(() => this.scrollToMiddle(400), 0);
    }
  }

  onSelect(e) {
    e.preventDefault();
    this.props.onSelect(this.props.id);
  }

  scrollToMiddle(duration) {
    const itemRect = this.element.getBoundingClientRect();
    this.context.scrollToMiddle(duration, itemRect.top, itemRect.height);
  }

  render() {
    const { id, active, children } = this.props;
    return (
      <li ref={(e) => { this.element = e; }} className={`ui-snap-listview-item${active ? ' ui-snap-listview-selected' : ''}`}>
        <a href={`#${id}`} onClick={this.onSelect} id={id}>{children}</a>
      </li>
    );
  }
}
