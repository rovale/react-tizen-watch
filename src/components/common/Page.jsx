import React from 'react';
import PropTypes from 'prop-types';
import scroll from 'smooth-move';

export class Page extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  static childContextTypes = {
    scrollToMiddle: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.element = null;
  }

  getChildContext() {
    return { scrollToMiddle: this.scrollToMiddle };
  }

  scrollToMiddle = (duration, itemTop, itemHeight) => {
    const scroller = this.element.getElementsByClassName('ui-scroller')[0];
    const scrollerHeight = scroller.getBoundingClientRect().height;
    const newScrollTop = (itemTop + scroller.scrollTop) - ((scrollerHeight - itemHeight) / 2);

    scroll(scroller, { x: 0, y: newScrollTop, duration });
  }

  render() {
    return (
      <div ref={(e) => { this.element = e; }} className="ui-page ui-page-active ui-scroll-on">
        <div className="ui-scroller ui-snap-container" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
          {this.props.children}
        </div>
      </div>
    );
  }
}

export const Header = ({ children }) => (
  <header className="ui-header">
    <h2 className="ui-title">{children}</h2>
  </header>
);

Header.propTypes = {
  children: PropTypes.string.isRequired,
};

export const Content = ({ children }) => {
  return (
    <div className="ui-content" ref={e => e && e.setAttribute('tizen-circular-scrollbar', '')}>
      {children}
    </div>);
};

Content.propTypes = {
  children: PropTypes.node.isRequired,
};
