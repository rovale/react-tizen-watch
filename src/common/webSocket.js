import io from 'socket.io-client';
import settings from './settings';
import * as action from '../actions/creators';

let socket;

export const initWebSocket = (store) => {
  const socketIoClient = window.testSocketIoClient || io;

  socket = socketIoClient(`${settings.protocol}://${settings.host}:${settings.port}/?username=${settings.userName}&password=${settings.password}`, {
    reconnection: true,
    reconnectionDelay: 1000,
    reconnectionDelayMax: 3000,
    timeout: 20000,
    forceNew: true,
  });

  socket.on('devices', (devices) => {
    store.dispatch(action.loadDevices(devices));
  });

  socket.on('pages', (pages) => {
    store.dispatch(action.loadPages(pages));
  });

  socket.on('deviceAttributeChanged', (event) => {
    if (event.deviceId !== 'syssensor' && event.deviceId !== 'temperature') {
      store.dispatch(action.changeDeviceAttribute(event));
    }
  });
};

export const toggleSwitch = (id) => {
  socket.emit('call', {
    id: 'callDeviceAction',
    action: 'callDeviceAction',
    params: {
      deviceId: id,
      actionName: 'toggle',
    },
  });
};

export const clickButton = (deviceId, buttonId) => {
  socket.emit('call', {
    id: 'callDeviceAction',
    action: 'callDeviceAction',
    params: {
      deviceId,
      buttonId,
      actionName: 'buttonPressed',
    },
  });
};
