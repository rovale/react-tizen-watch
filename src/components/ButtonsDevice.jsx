import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { List } from './common/List';

import * as action from '../actions/creators';

let ButtonsDevice = ({ device, activeButtonId, onActivateButton, onClickButton }) => {
  const buttons = device.config.buttons;
  const activeOptionId = activeButtonId || buttons[0].id;

  return (<List
    options={buttons.map(b => ({ id: b.id, title: b.text }))}
    activeOptionId={activeOptionId}
    onActivateOption={onActivateButton}
    onSelectOption={onClickButton}
  />);
};

ButtonsDevice.propTypes = {
  device: PropTypes.object.isRequired,
  activeButtonId: PropTypes.string,
  onActivateButton: PropTypes.func.isRequired,
  onClickButton: PropTypes.func.isRequired,
};

ButtonsDevice.defaultProps = {
  activeButtonId: null,
};

const mapStateToProps = state => ({
  activeButtonId: state.ui.activeButtonId,
});

const mapDispatchToProps = dispatch => ({
  onActivateButton: id => dispatch(action.activateButton(id)),
  onClickButton: id => dispatch(action.clickButton(id)),
});

ButtonsDevice = connect(mapStateToProps, mapDispatchToProps)(ButtonsDevice);

export default ButtonsDevice;
