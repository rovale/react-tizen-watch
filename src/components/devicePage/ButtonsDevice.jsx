import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { List } from '../common/List';

import * as action from '../../actions/creators';

const ButtonsDevice = ({
  device, activeButtonId, onActivateButton, onClickButton,
}) => {
  const { buttons } = device.config;
  const activeOptionId = activeButtonId || buttons[0].id;

  const getListItemContent = button => (
    <div>
      <div className="vertical-align-middle">
        <input
          type="radio"
          readOnly
          checked={button.id === device.attributes[0].value}
        />
      </div>
      <div className="vertical-align-middle">
        &nbsp;{button.text}
      </div>
    </div>);

  return (<List
    options={buttons.map(b => ({
      id: b.id,
      content: getListItemContent(b),
    }))}
    activeOptionId={activeOptionId}
    onActivateOption={onActivateButton}
    onSelectOption={onClickButton}
  />);
};

ButtonsDevice.propTypes = {
  device: PropTypes.shape({
    attributes: PropTypes.array.isRequired,
    config: PropTypes.object.isRequired,
  }).isRequired,
  activeButtonId: PropTypes.string,
  onActivateButton: PropTypes.func.isRequired,
  onClickButton: PropTypes.func.isRequired,
};

ButtonsDevice.defaultProps = {
  activeButtonId: null,
};

const mapStateToProps = state => ({
  activeButtonId: state.ui.activeButtonId,
});

const mapDispatchToProps = dispatch => ({
  onActivateButton: id => dispatch(action.activateButton(id)),
  onClickButton: id => dispatch(action.clickButton(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ButtonsDevice);
