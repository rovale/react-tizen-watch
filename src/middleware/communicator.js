import * as actionType from '../actions/types';
import { initWebSocket, toggleSwitch, clickButton } from '../common/webSocket';

// eslint-disable-next-line arrow-parens
const communicator = store => next => action => {
  switch (action.type) {
    case actionType.INITIALIZE_APP:
      initWebSocket(store);
      break;
    case actionType.TOGGLE_SWITCH:
      toggleSwitch(action.payload.id);
      break;
    case actionType.CLICK_BUTTON: {
      const deviceId = store.getState().ui.selectedDeviceId;
      clickButton(deviceId, action.payload.id);
      break;
    }
    default:
      break;
  }

  return next(action);
};

export default communicator;
