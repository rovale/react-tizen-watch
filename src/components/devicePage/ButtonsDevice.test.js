import React from 'react';
import { shallow } from 'enzyme';

import { createStore, applyMiddleware, compose } from 'redux';
import reducer from '../../reducers';
import communicator from '../../middleware/communicator';
import * as action from '../../actions/creators';

import ButtonsDevice from './ButtonsDevice';
import { List, Item } from './../common/List';

const devices = [
  {
    id: 'pick-a-color-radio',
    name: 'Pick a color',
    template: 'buttons',
    attributes: [
      {
        name: 'button',
        value: 'blue',
      },
    ],
    config: {
      buttons: [
        { id: 'red', text: 'Red' },
        { id: 'green', text: 'Green' },
        { id: 'blue', text: 'Blue' },
      ],
    },
  },
];

describe('The ButtonsDevice component', () => {
  let store;
  let device;
  let socketIoEmitSpy;

  beforeEach(() => {
    socketIoEmitSpy = jest.fn();
    window.testSocketIoClient = () => ({ emit: socketIoEmitSpy, on: jest.fn() });

    const enhancer = compose(applyMiddleware(communicator));
    store = createStore(reducer, enhancer);
    store.dispatch(action.initializeApp());
    store.dispatch(action.loadDevices(devices));
    [device] = store.getState().devices;
  });

  // TODO: enable when hooks are no longer alpha
  // test('it should show a list with the buttons as options', () => {
  //   const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();

  //   const listContent = buttonsDevice.dive();
  //   expect(listContent).toMatchSnapshot();
  //   expect(listContent.find(Item)).toHaveLength(3);

  //   const checkBox = listContent.find('Item[id="blue"] input');
  //   expect(checkBox.prop('checked')).toBeTruthy();
  // });

  // TODO: enable when hooks are no longer alpha
  // test('it should reflect a device attribute change', () => {
  //   store.dispatch(action.changeDeviceAttribute({
  //     deviceId: 'pick-a-color-radio',
  //     attributeName: 'button',
  //     value: 'red',
  //   }));

  //   [device] = store.getState().devices;
  //   const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();

  //   const listContent = buttonsDevice.dive();
  //   const checkBox = listContent.find('Item[id="red"] input');
  //   expect(checkBox.prop('checked')).toBeTruthy();
  // });

  test('it should mark the first button as active one', () => {
    const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();
    const list = buttonsDevice.find(List);

    expect(list.prop('activeOptionId')).toBe('red');
  });

  test('it should mark the activated button as active button', () => {
    store.dispatch(action.activateButton('green'));
    const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();
    const list = buttonsDevice.find(List);

    expect(list.prop('activeOptionId')).toBe('green');
  });

  test('it should activate the button when the option is activated in the list', () => {
    const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();
    const list = buttonsDevice.find(List);

    list.prop('onActivateOption')('blue');

    expect(store.getState().ui.activeButtonId).toBe('blue');
  });

  // TODO: enable when hooks are no longer alpha
  // test('it should activate the button when the button is selected', () => {
  //   const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();
  //   const list = buttonsDevice.find(List);
  //   const listContent = list.dive();
  //   const listItem = listContent.find('Item[id="blue"]');

  //   // TODO: make onSelect work without specifying the id.
  //   listItem.prop('onSelect')('blue');

  //   expect(store.getState().ui.activeButtonId).toBe('blue');
  // });

  // TODO: enable when hooks are no longer alpha
  // test('it should emit the expected command when the button is selected', () => {
  //   const buttonsDevice = shallow((<ButtonsDevice store={store} device={device} />)).dive();
  //   const list = buttonsDevice.find(List);
  //   const listContent = list.dive();
  //   const listItem = listContent.find('Item[id="blue"]');

  //   // TODO: make onSelect work without specifying the id.
  //   listItem.prop('onSelect')('blue');

  //   expect(socketIoEmitSpy.mock.calls).toMatchSnapshot();
  // });
});
