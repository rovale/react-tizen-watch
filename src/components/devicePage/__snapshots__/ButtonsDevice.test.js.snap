// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ButtonsDevice component it should emit the expected command when the button is selected 1`] = `
Array [
  Array [
    "call",
    Object {
      "action": "callDeviceAction",
      "id": "callDeviceAction",
      "params": Object {
        "actionName": "buttonPressed",
        "buttonId": "blue",
        "deviceId": null,
      },
    },
  ],
]
`;

exports[`The ButtonsDevice component it should show a list with the buttons as options 1`] = `
<ul
  className="ui-listview ui-snap-listview"
>
  <Item
    active={true}
    id="red"
    key="red"
    onSelect={[Function]}
  >
    <div>
      <div
        className="vertical-align-middle"
      >
        <input
          checked={false}
          readOnly={true}
          type="radio"
        />
      </div>
      <div
        className="vertical-align-middle"
      >
         
        Red
      </div>
    </div>
  </Item>
  <Item
    active={false}
    id="green"
    key="green"
    onSelect={[Function]}
  >
    <div>
      <div
        className="vertical-align-middle"
      >
        <input
          checked={false}
          readOnly={true}
          type="radio"
        />
      </div>
      <div
        className="vertical-align-middle"
      >
         
        Green
      </div>
    </div>
  </Item>
  <Item
    active={false}
    id="blue"
    key="blue"
    onSelect={[Function]}
  >
    <div>
      <div
        className="vertical-align-middle"
      >
        <input
          checked={true}
          readOnly={true}
          type="radio"
        />
      </div>
      <div
        className="vertical-align-middle"
      >
         
        Blue
      </div>
    </div>
  </Item>
</ul>
`;
